src/gwt/src/org/rstudio/studio/client/workbench/views/source/SourceColumn.java

```diff
@@ -0,0 +1,1215 @@
+      if (display_.getTabCount() > 0 && display_.getActiveTabIndex() >= 0)
+         editors_.get(index).onInitiallyLoaded();
+   }
+
+   /**
+    * @param isNewTabPending True if a new tab is about to be created. (If
+    *    false and there are no tabs already, then a new source doc might
+    *    be created to make sure we don't end up with a source pane showing
+    *    with no tabs in it.)
+    */
+   public void ensureVisible(boolean isNewTabPending)
+   {
+      newTabPending_++;
+      try
+      {
+         display_.ensureVisible();
+      }
+      finally
+      {
+         newTabPending_--;
+      }
+   }
+
+   // when tabs have been reordered in the session, the physical layout of the
+   // tabs doesn't match the logical order of editors_. it's occasionally
+   // necessary to get or set the tabs by their physical order.
+   public int getPhysicalTabIndex()
+   {
+      int idx = display_.getActiveTabIndex();
+      if (idx < tabOrder_.size())
+      {
+         idx = tabOrder_.indexOf(idx);
+      }
+      return idx;
+   }
+
+   public void setPhysicalTabIndex(int idx)
+   {
+      if (idx < tabOrder_.size())
+      {
+         idx = tabOrder_.get(idx);
+      }
+      display_.selectTab(idx);
+   }
+
+   private void onActivate(EditingTarget target)
+   {
+       // return if we're already set properly
+       if (activeEditor_ != null && activeEditor_ == target)
+          return;
+
+       // deactivate prior active editor
+       if (activeEditor_ != null)
+          activeEditor_.onDeactivate();
+
+       // set and active editor
+       activeEditor_ = target;
+       if (activeEditor_ != null)
+          activeEditor_.onActivate();
+       manageCommands();
+   }
+
+   void setActiveEditor()
+   {
+       if (activeEditor_ == null &&
+           editors_.size() > display_.getActiveTabIndex())
+          onActivate(editors_.get(display_.getActiveTabIndex()));
+   }
+
+   EditingTarget setActiveEditor(String docId)
+   {
+      if (StringUtil.isNullOrEmpty(docId) &&
+          activeEditor_ != null)
+      {
+         activeEditor_.onDeactivate();
+         activeEditor_ = null;
+         return null;
+      }
+
+      for (EditingTarget target : editors_)
+      {
+         if (target.getId().equals(docId))
+         {
+             onActivate(target);
+            return target;
+         }
+      }
+      return null;
+   }
+
+   public void setActiveEditor(EditingTarget target)
+   {
+      // This should never happen
+      if (!editors_.contains(target))
+      {
+         Debug.logWarning("Attempting to set active editor to an unknown target.");
+         return;
+      }
+      onActivate(target);
+   }
+
+   private void syncTabOrder()
+   {
+      // ensure the tab order is synced to the list of editors
+      for (int i = tabOrder_.size(); i < editors_.size(); i++)
+      {
+         tabOrder_.add(i);
+      }
+      for (int i = editors_.size(); i < tabOrder_.size(); i++)
+      {
+         tabOrder_.remove(i);
```

Comment:
Is this right? At first glance seems like this won't work but I might be misunderstanding.

Reply:
Yes, because each time we are only modifying `tabOrder`. This is taken from today's Source.java https://github.com/rstudio/rstudio/blob/master/src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java#L3818-L3829